### PR TITLE
[api] refactor gas estimation code out of core service 

### LINKED ISSUE #2905: refactor gas estimation code out of core service
https://github.com/iotexproject/iotex-core/blob/960f334eff0843e94056f4fde8a6f4e545d79267/api/coreservice.go#L420-L423

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in 960f334eff0843e94056f4fde8a6f4e545d79267 when #2888 was merged. cc @Liuhaai.

### PR DESCRIPTION
**Pull request by Liuhaai**
fix #2905

### CODE REVIEW

File: api/grpcserver.go [hunk 1]
```diff
@@ -265,8 +265,37 @@ func (svr *GRPCServer) EstimateGasForAction(ctx context.Context, in *iotexapi.Es
...
 // EstimateActionGasConsumption estimate gas consume for action without signature
-func (svr *GRPCServer) EstimateActionGasConsumption(ctx context.Context, in *iotexapi.EstimateActionGasConsumptionRequest) (respone *iotexapi.EstimateActionGasConsumptionResponse, err error) {
-	ret, err := svr.coreService.EstimateActionGasConsumption(ctx, in)
+func (svr *GRPCServer) EstimateActionGasConsumption(ctx context.Context, in *iotexapi.EstimateActionGasConsumptionRequest) (*iotexapi.EstimateActionGasConsumptionResponse, error) {
+	var (
+		ret uint64
+		err error
+	)
+	switch {
+	case in.GetExecution() != nil:
```
Comment (CoderZhi) [inline api/grpcserver.go:274]: early return
```
    if exec := in.GetExecution(); exec != nil {
        ret, err := src.coreService.EstimateExecutionGasConsumption(ctx, exec, in.GetCallerAddress())
        if err != nil {
             return nil, err
        }
        return ...
    }
    var intrinsicGas, payloadGas, payloadSize uint64
    ......
    // default: return err
    return return &iotexapi.EstimateActionGasConsumptionResponse{
        Gas: src.coreService.CalculateGasConsumption(intrinsicGas, payloadGas, payloadSize),
    }, nil
```

File: api/web3server.go [hunk 2]
```diff
@@ -373,17 +373,13 @@ func (svr *Web3Server) getTransactionCount(in interface{}) (interface{}, error)
...
 
 func (svr *Web3Server) call(in interface{}) (interface{}, error) {
-	from, to, gasLimit, value, data, err := parseCallObject(in)
+	callerAddr, to, gasLimit, value, data, err := parseCallObject(in)
 	if err != nil {
 		return nil, err
 	}
 	if to == "io1k8uw2hrlvnfq8s2qpwwc24ws2ru54heenx8chr" {
 		return nil, nil
 	}
```
Comment (dustinxie) [inline api/web3server.go:382]: why check "io1k8uw2hrlvnfq8s2qpwwc24ws2ru54heenx8chr"?

Comment (dustinxie) [reply api/web3server.go:382]: pls add a comment that this is the special balance checker contract (which metamask will check periodically)

**Review by dustinxie — APPROVED** [review]

**Review by CoderZhi — APPROVED** [review]

File: api/web3server.go [hunk 3]
```diff
@@ -450,7 +447,8 @@ func (svr *Web3Server) estimateGas(in interface{}) (interface{}, error) {
 		}
 	case false:
 		// transfer
-		estimatedGas = uint64(len(data))*action.TransferPayloadGas + action.TransferBaseIntrinsicGas
+		estimatedGas = svr.coreService.CalculateGasConsumption(action.TransferBaseIntrinsicGas,
```
Comment (millken) [inline api/web3server.go:450]: can use if instead of switch

**Pull request merged**